src/ReactPlaceholder.tsx

```diff
@@ -1,128 +1,142 @@
+import { joinClassNames } from './utils';
 
 export type CommonProps = {
-  children: React.ReactNode,
+  children: React.ReactElement | null;
   /** pass `true` when the content is ready and `false` when it's loading */
-  ready: boolean,
+  ready: boolean;
   /** delay in millis to wait when passing from ready to NOT ready */
-  delay?: number,
+  delay?: number;
   /** if true, the placeholder will never be rendered again once ready becomes true, even if it becomes false again */
-  firstLaunchOnly?: boolean,
-  className?: string,
-  style?: React.CSSProperties
-}
-
-export type Props = (CommonProps & {
-  /** type of placeholder to use */
-  type: 'text' | 'media' | 'textRow' | 'rect' | 'round',
-  /** number of rows displayed in 'media' and 'text' placeholders */
-  rows?: number,
-  /** color of the placeholder */
-  color?: string,
-  /** pass true to show a nice loading animation on the placeholder */
-  showLoadingAnimation?: boolean,
-  customPlaceholder?: undefined
-}) | (CommonProps & {
+  firstLaunchOnly?: boolean;
+  className?: string;
+  style?: React.CSSProperties;
+};
+
+export type PlaceholderProps = CommonProps & {
+  // we have a default color, so we can set this as optional
+  color?: string;
+  // we have a default color, so we can set this as optional
+  rows?: number;
+  showLoadingAnimation?: boolean;
+  customPlaceholder?: undefined;
+};
+
+export type CustomPlaceholderProps = CommonProps & {
   /** pass any renderable content to be used as placeholder instead of the built-in ones */
-  customPlaceholder?: React.ReactNode | React.ReactElement<{ [k: string]: any }>,
-  type?: undefined,
-  rows?: undefined,
-  color?: undefined,
-  showLoadingAnimation?: undefined
-})
-
-export default class ReactPlaceholder extends React.Component<Props> {
-
-  static propTypes = {
-    children: PropTypes.oneOfType([
-      PropTypes.node,
-      PropTypes.element
-    ]).isRequired,
-    delay: PropTypes.number,
-    ready: PropTypes.bool.isRequired,
-    firstLaunchOnly: PropTypes.bool,
-    type: PropTypes.oneOf(['text', 'media', 'textRow', 'rect', 'round']),
-    rows: PropTypes.number,
-    color: PropTypes.string,
-    showLoadingAnimation: PropTypes.bool,
-    customPlaceholder: PropTypes.oneOfType([
-      PropTypes.node,
-      PropTypes.element
-    ]),
-    className: PropTypes.string,
-    style: PropTypes.object
-  }
-
-  static defaultProps = {
-    delay: 0,
-    type: 'text',
-    color: '#CDCDCD'
-  }
-
-  state = {
-    ready: this.props.ready
-  }
-
-  timeout?: number;
-
-  getFiller = () => {
-    const {
-      firstLaunchOnly, children, ready, className, // eslint-disable-line no-unused-vars
-      type, customPlaceholder, showLoadingAnimation, ...rest
-    } = this.props;
-
-    const classes = showLoadingAnimation ?
-      ['show-loading-animation', className].filter(c => c).join(' ') :
-      className;
+  customPlaceholder: React.ReactElement<{ [k: string]: any }> | null;
```

Comment:
why removing the optionality? In this way the prop should always be specified to `={null}` when not needed, right?

Reply:
Oops, yeah, good point. Even though the props are more isomorphic, we still want to allow no type of custom placeholder to be passed in and have the component use the default props.